JWT support for EL engine calls

arhamj opened the pull request:
<!-- Thanks for sending a pull request! Please check out our contribution guidelines: -->
<!-- https://github.com/ConsenSys/teku/blob/master/CONTRIBUTING.md -->

## PR Description

Pending

- [x] Generating secret key if not present
- [x] Adding UTs

Others
1. Went with adding caching of the Jwt token
2. Retries on failure is already handled

## Fixed Issue(s)
<!-- Please link to fixed issue(s) here using format: fixes #<issue number> -->
<!-- Example: "fixes #2" -->

fixes #4958

## Documentation

- [x] I thought about documentation and added the `documentation` label to this PR if updates are required.

## Changelog

- [x] I thought about adding a changelog entry, and added one if I deemed necessary.

arhamj committed:
feat(auth): added base code for jwt secret cli flag and http interceptor

arhamj committed:
feat(jwt): added okhttp auth header interceptor

arhamj committed:
style(reformat): code reformat as per style guide

arhamj committed:
feat(pojo): updated to use standard java pojos for keys instead of strings

arhamj committed:
refactor(time-provider): using time provider interface to help with tests

arhamj committed:
Merge branch 'master' into jwt-for-el-calls

ajsutton commented on ethereum/executionlayer/src/main/java/tech/pegasys/teku/ethereum/executionlayer/client/auth/JwtAuthInterceptor.java:
```diff
@@ -0,0 +1,48 @@
+      return chain.proceed(chain.request());
+    }
+    final Token token = optionalToken.get();
+    final String authHeader = String.format("%s", token.getJwtToken());
```
Couldn't this just be 
```suggestion
    final String authHeader = token.getJwtToken().toString();
```

arhamj replied:
Was not sure if we were going to have a prefix to the token, so just kept it this way. Will update

ajsutton commented on ethereum/executionlayer/src/main/java/tech/pegasys/teku/ethereum/executionlayer/client/auth/SafeTokenProvider.java:
```diff
@@ -0,0 +1,53 @@
+
+public class SafeTokenProvider {
+  private final ReentrantLock lock = new ReentrantLock();
+  private Token token;
```
Probably makes sense to make this `Optional<Token>` to make it clear that it may not be set.

ajsutton commented on ethereum/executionlayer/src/main/java/tech/pegasys/teku/ethereum/executionlayer/client/auth/SafeTokenProvider.java:
```diff
@@ -0,0 +1,53 @@
+    if (possibleToken.isEmpty()) {
+      return possibleToken;
+    }
+    token = possibleToken.get();
```
Not entirely sure, but I think if the provider doesn't return a token we probably should clear the cached token.  If `token` was an `Optional` then we could just skip the `isEmpty()` check here and always set `token`.

ajsutton commented on ethereum/executionlayer/src/main/java/tech/pegasys/teku/ethereum/executionlayer/client/auth/TokenProvider.java:
```diff
@@ -0,0 +1,38 @@
+  }
+
+  public Optional<Token> token(final Date instant) {
+    final long expiresInMilliseconds = jwtConfig.getExpiresInSeconds() * 1000;
```
```suggestion
    final long expiresInMilliseconds = TimeUnit.SECONDS.toMillis(jwtConfig.getExpiresInSeconds());
```

ajsutton commented on gradle/versions.gradle:
```diff
@@ -162,5 +162,6 @@ dependencyManagement {
 
     dependency 'org.jupnp:org.jupnp.support:2.6.0'
     dependency 'org.jupnp:org.jupnp:2.6.0'
+    dependency 'io.jsonwebtoken:jjwt:0.9.1'
```
This seems to be an old version - there are tags for 0.11.2.

Also, any particular reason to use this JWT library? I don't have a particular preference but it's slightly concerning it hasn't reached it's 1.0 release and there seem to be a number of potential alternatives that may be more mature. For example auth0 have a library (https://github.com/auth0/java-jwt)

arhamj replied:
[mvn repository](https://mvnrepository.com/artifact/io.jsonwebtoken/jjwt) made me believe that it was the latest version. I used this library as it had the highest number of stars. But looking at it now it does have a lot more reported issues compared to `java-jwt`. I'll check out a few libraries and update

ajsutton commented on teku/src/main/java/tech/pegasys/teku/cli/options/ExecutionEngineOptions.java:
```diff
@@ -36,8 +36,20 @@
 
+  @Option(
+      hidden = true,
+      names = {"--Xjwt-secret"},
```
For consistency we probably should call this:
```suggestion
      names = {"--Xee-jwt-secret"},
```

arhamj committed:
refactor(review): updated to address comments

arhamj committed:
refactor(review): updated jjwt library version

arhamj committed:
feat(auth): removed auth header string formatting

ajsutton commented on ethereum/executionlayer/src/main/java/tech/pegasys/teku/ethereum/executionlayer/client/auth/SafeTokenProvider.java:
```diff
@@ -0,0 +1,50 @@
+    this.tokenProvider = tokenProvider;
+  }
+
+  public Optional<Token> token(final UInt64 instantInMillis) {
```
Small detail but it would probably be simpler to just mark this method as `synchronized` rather than using a `ReentrantLock`. We don't need any of the monitoring features of `ReentrantLock`.

arhamj replied:
makes sense. will update and add tests today. Will also be updating to use the Auth0 JWT library

ajsutton reviewed (commented):
Looking good. The build is failing because error prone wants you to use `Instant` rather than `Date`. I was wondering if that would be possible as well because things like time comparison and adding works better with `Instant`.  I'm just not sure if using `Instant` makes it too hard to work with the jwt library if it's expecting a `Date` though. If it does we could suppress the warning.

arhamj committed:
test(ut): added jwt auth tests

arhamj committed:
Merge branch 'master' into jwt-for-el-calls

arhamj committed:
style(rename): update variable name to indicate time unit

ajsutton commented on ethereum/executionlayer/build.gradle:
```diff
@@ -10,6 +10,9 @@ dependencies {
   implementation 'org.apache.tuweni:tuweni-units'
+  implementation 'io.jsonwebtoken:jjwt-api'
+  implementation 'io.jsonwebtoken:jjwt-impl'
+  implementation 'io.jsonwebtoken:jjwt-jackson'
```
I think these two should be `runtimeOnly` so they aren't on the compile class path and we're forced to only use the stable API.

ajsutton commented on ethereum/executionlayer/src/main/java/tech/pegasys/teku/ethereum/executionlayer/client/auth/JwtSecretKeyLoader.java:
```diff
@@ -0,0 +1,56 @@
+
+  private Key generateNewSecret() {
+    LOG.info("generating new execution engine JWT secret");
+    final Key key = MacProvider.generateKey(SignatureAlgorithm.HS256);
```
I think the public API for this is:
```suggestion
    final Key key = Keys.secretKeyFor(SignatureAlgorithm.HS256);
```

ajsutton commented on ethereum/executionlayer/src/main/java/tech/pegasys/teku/ethereum/executionlayer/client/auth/SafeTokenProvider.java:
```diff
@@ -0,0 +1,43 @@
+    if (token.isEmpty()) {
+      return refreshTokenFromSource(instantInMillis);
+    }
+    if (token.get().isAvailableAt(instantInMillis)) {
```
I think we can simplify this to:

```suggestion
    if (this.token.map(token -> token.isAvailableAt(instantInMillis)).orElse(false)) {
      return token;
    }
```

So return the token if it exists and is available at the specified time.

ajsutton commented on ethereum/executionlayer/src/test/java/tech/pegasys/teku/ethereum/executionlayer/client/auth/JwtSecretKeyLoaderTest.java:
```diff
@@ -0,0 +1,51 @@
+  void testGetSecretKey_KeyGeneration() {
+    final JwtSecretKeyLoader keyLoader = new JwtSecretKeyLoader(Optional.empty());
+    final Key loadedSecret = keyLoader.getSecretKey();
+    assertThat(Objects.isNull(loadedSecret)).isFalse();
```
```suggestion
    assertThat(loadedSecret).isNotNull();
```

ajsutton commented on ethereum/executionlayer/src/test/java/tech/pegasys/teku/ethereum/executionlayer/client/auth/JwtSecretKeyLoaderTest.java:
```diff
@@ -0,0 +1,51 @@
+    final JwtSecretKeyLoader keyLoader = new JwtSecretKeyLoader(Optional.empty());
+    final Key loadedSecret = keyLoader.getSecretKey();
+    assertThat(Objects.isNull(loadedSecret)).isFalse();
+    assertThat(StringUtils.isNotEmpty(Bytes.wrap(loadedSecret.getEncoded()).toHexString()))
```
```suggestion
    assertThat(Bytes.wrap(loadedSecret.getEncoded()).toHexString())).isNotEmpty();
```
I think that's the right method

ajsutton commented on ethereum/executionlayer/src/test/java/tech/pegasys/teku/ethereum/executionlayer/client/auth/SafeTokenProviderTest.java:
```diff
@@ -0,0 +1,96 @@
+    assertThat(
+            validateJwtTokenAtInstant(
+                jwtSecretKey, originalToken.get().getJwtToken(), timeInMillis))
+        .isTrue();
```
I'd probably move the assertions into `validateJwtTokenAtInstance` so that it can use the richer methods `assertThat` provides rather than returning a boolean.  That way we'll get much better error messages if the test fails instead of just "expected true but was false".

ajsutton commented on ethereum/executionlayer/src/test/java/tech/pegasys/teku/ethereum/executionlayer/client/auth/SafeTokenProviderTest.java:
```diff
@@ -0,0 +1,96 @@
+            validateJwtTokenAtInstant(jwtSecretKey, updatedToken.get().getJwtToken(), timeInMillis))
+        .isTrue();
+
+    assertThat(originalToken.equals(updatedToken)).isTrue();
```
```suggestion
    assertThat(originalToken).isEqualTo(updatedToken);
```

ajsutton commented on services/executionengine/src/main/java/tech/pegasys/teku/services/executionengine/ExecutionEngineConfiguration.java:
```diff
@@ -81,5 +93,11 @@ public Builder specProvider(final Spec spec) {
+
+    public Builder jwtSecretFile(final String jwtSecretFile) {
+      checkNotNull(jwtSecretFile);
+      this.jwtSecretFile = Optional.of(jwtSecretFile).filter(f -> !f.isBlank());
```
This will need to support `jwtSecretFile` being `null` since that's the default value in the cli definition. This is what's causing most of the test failures. And we can simplify a little by using the isNotBlank method from commons-lang3. So probably:
```suggestion
      this.jwtSecretFile = Optional.ofNullable(jwtSecretFile).filter(StringUtils::isNotBlank);
```

ajsutton commented on services/executionengine/src/main/java/tech/pegasys/teku/services/executionengine/ExecutionEngineConfiguration.java:
```diff
@@ -59,12 +70,13 @@ public Version getVersion() {
     private Spec spec;
     private Optional<String> endpoint = Optional.empty();
     private Version version = Version.DEFAULT_VERSION;
+    private Optional<String> jwtSecretFile;
```
Need to provide a default value here:
```suggestion
    private Optional<String> jwtSecretFile = Optional.empty();
```

ajsutton reviewed (commented):
Looking good, thanks for your work on this.

arhamj committed:
fix(review): fixed failing tests, updated as per comments

arhamj committed:
feat(jwt): writing jwt secret to beacon data dir

ajsutton committed:
Merge branch 'master' of github.com:ConsenSys/teku into jwt-for-el-calls

Avoid adding authentication to kintsugi and kilnv1 specs. Auth only applies to kilnv2 onwards.

ajsutton committed:
Minor tidy ups:
Use assertj better in tests.
Improve error messages when the jwt-secret file can't be found.

ajsutton reviewed (approved):
LGTM. I've pushed up a couple of small changes so that we only add authentication when the kilnv2 ee version is used, and to ensure we get a clear error message if the jwt-secret file can't be read at startup.

Thanks for getting this done, it's really great work.

At this stage I think we're spec compliant and since kilnv2 isn't the default version yet I'm ok with merging it in. The next step will be testing against EL clients that support authentication to ensure it really is compatible and then working through any issue we find.
